pallets/automation-time/src/lib.rs

```diff
@@ -1343,27 +1325,74 @@ pub mod pallet {
 			Ok(())
 		}
 
-		/// Reschedules an existing task for a given number of execution times
-		fn reschedule_existing_task(
+		fn reschedule_or_remove_task(
 			task_id: TaskId<T>,
-			task: &TaskOf<T>,
-			execution_times: Vec<UnixTime>,
-		) -> Result<(), DispatchError> {
-			let new_executions = execution_times.len().try_into().unwrap();
+			task: &mut TaskOf<T>,
+			dispatch_error: Option<DispatchError>,
+		) {
+			let remove_on_err = |task: &TaskOf<T>, err: DispatchError| {
+				Self::deposit_event(Event::<T>::TaskNotRescheduled {
+					who: task.owner_id.clone(),
+					task_id,
+					error: err,
+				});
+				AccountTasks::<T>::remove(task.owner_id.clone(), task_id);
+			};
 
-			let fee = Self::calculate_execution_fee(&task.action, new_executions)?;
-			T::FeeHandler::can_pay_fee(&task.owner_id, fee.clone())
-				.map_err(|_| Error::<T>::InsufficientBalance)?;
+			if let Some(e) = dispatch_error {
+				remove_on_err(task, e);
+			} else {
+				let owner_id = task.owner_id.clone();
+				match Self::reschedule_existing_task(task_id, task) {
+					Ok(_) => {
+						Self::deposit_event(Event::<T>::TaskRescheduled { who: owner_id, task_id });
+					},
+					Err(e) => {
+						remove_on_err(task, e);
+					},
+				};
+			}
+		}
 
-			Self::insert_scheduled_tasks(task_id, task, execution_times.clone())?;
+		fn reschedule_existing_task(task_id: TaskId<T>, task: &mut TaskOf<T>) -> DispatchResult {
+			match task.schedule {
+				Schedule::Recurring { ref mut next_execution_time, frequency } => {
+					let new_execution_time = next_execution_time
```

Comment:
shouldn't this logic also live on the struct?

Reply:
I think that's reasonable. I didn't originally include it because it requires passing the `task` into the method and dependency felt a little too circular. I think I'd want to split this across the Task and Schedule struct to make that a bit cleaner. Will think on this.

Reply:
Splitting this suggestion into a new pr so we can discuss the pros/cons in isolation.